core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       auditContext.setSrcInode(inode);
 
       List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
-        try {
-          mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
-          throw e;
-        }
-        for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-          child.lockReadAndCheckParent(inode);
+      listStatusRecursive(ret, inodePath, auditContext, listStatusOptions.isRecursive());
+
+      auditContext.setSucceeded(true);
+      Metrics.FILE_INFOS_GOT.inc();
+      return ret;
+    }
+  }
+
+  private void listStatusRecursive(List<FileInfo> statusList, LockedInodePath originalInodePath,
+                                   AuditContext auditContext, boolean recursive)
+      throws UnavailableException, FileDoesNotExistException,
+      AccessControlException, InvalidPathException {
+    Queue<LockedInodePath> queue = new LinkedList<>();
+    queue.add(originalInodePath);
+
+    try (InodeLockList lockList = new InodeLockList()) {
+      while (!queue.isEmpty()) {
+        LockedInodePath currInodePath = queue.remove();
+        Inode<?> inode = currInodePath.getInode();
```

Comment:
this only works for a full path. Are these guaranteed to be full? Are we expecting that the target inode always exists?

Reply:
For the `TempInodePathForDescendant` kind, this does not really apply. For the very first node that we deal with, it is guaranteed to be full. since we call `ensureFullPathAndUpdateCache` before calling `listStatusInternal`, but maybe we should check again?